stan/math/prim/mat/prob/normal_id_glm_lpdf.hpp

```diff
@@ -92,20 +100,101 @@ return_type_t<T_y, T_x, T_alpha, T_beta, T_scale> normal_id_glm_lpdf(
   const auto &y_val_vec = as_column_vector_or_scalar(y_val);
 
   T_scale_val inv_sigma = 1 / as_array_or_scalar(sigma_val_vec);
+  Matrix<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma_mat(N);
+  auto y_minus_mu_over_sigma = y_minus_mu_over_sigma_mat.array();
+
+  operands_and_partials<T_y, T_x, T_alpha, T_beta, T_scale> ops_partials(
+      y, x, alpha, beta, sigma);
+
+  double y_minus_mu_over_sigma_squared_sum;  // the most efficient way to
+  // calculate this depends on
+  // template parameters
 
-  Array<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma
-      = x_val * beta_val_vec;
+#ifdef STAN_OPENCL
+  const int local_size
+      = opencl_kernels::normal_id_glm.make_functor.get_opts().at("LOCAL_SIZE_");
+  const int wgs = (N + local_size - 1) / local_size;
+
+  const matrix_cl<double> y_cl = matrix_cl<double>::constant(y_val_vec);
+  const matrix_cl<double> x_cl = matrix_cl<double>::constant(x_val);
+  matrix_cl<double> beta_cl(beta_val_vec);
+  matrix_cl<double> alpha_cl(alpha_val_vec);
+  matrix_cl<double> sigma_cl(sigma_val_vec);
+
+  const bool need_mu_derivative
+      = !is_constant_all<T_y, T_x, T_beta, T_alpha>::value;
+  matrix_cl<double> mu_derivative_cl(need_mu_derivative ? N : 0, 1);
+  const bool need_mu_derivative_sum
+      = !is_constant_all<T_alpha>::value && !is_vector<T_alpha>::value;
+  matrix_cl<double> mu_derivative_sum_cl(need_mu_derivative_sum ? wgs : 0, 1);
+  matrix_cl<double> y_minus_mu_over_sigma_squared_sum_cl(wgs, 1);
+  const bool need_sigma_derivative
+      = !is_constant_all<T_scale>::value && is_vector<T_scale>::value;
+  matrix_cl<double> sigma_derivative_cl(need_sigma_derivative ? N : 0, 1);
+  const bool need_log_sigma_sum
+      = include_summand<propto, T_scale>::value && is_vector<T_scale>::value;
```

Comment:
Can you pull all of these out of this ifdef and reuse them in the #else side?

Reply:
Also may be good to write a `is_not_constant<>` type trait

Reply:
You mean in CPU code? The problem is they are not exactley the same as conditions for branches on CPU side. I sure can reuse them for those branches in GPU version. I will see what I can do for CPU side.

Reply:
Actually I could only reuse `mu_derivative_sum_cl` on GPU side and none on CPU. Other `need_`... variables are only used for kernel arguments and not reused later.

Reply:
> Also may be good to write a is_not_constant<> type trait

I don't think we need a separate type trait for every thing we use - only of things that are tedious to write or complicated to understand. A simple `!` is neither IMHO. Anyway, the new trait would have to be named `is_not_constant_any`, which is longer than `!` ;).

Reply:
lol alright fair

Reply:
This was a topic of discussion already, we opted with going with ! : https://github.com/stan-dev/math/pull/1234#issuecomment-504744484